### PR TITLE
fix: Deduplicate namespace terms for ignore, ignore_linear, and keep when saving to model and printing out in command line.

### LINKED ISSUE #3335: Ignored namespace keeps duplicating in vw output when using ignore_linear with save_resume.
#### Describe the bug
When using both --ignore_linear and --save_resume, there are duplicated ignored namespace letters in vw output, see below:
"ignoring linear terms for namespaces beginning with: T T T T T T T T T"

~~~
...

$ cat xai | ./vw --ignore_linear T --save_resume -i weights -f weights
ignoring linear terms for namespaces beginning with: T T T T T T T T T
...

$ cat xaj | ./vw --ignore_linear T --save_resume -i weights -f weights
ignoring linear terms for namespaces beginning with: T T T T T T T T T T
...
~~~

#### To Reproduce
1. `split -l 100 train-sets/0002.dat`
2. run vw with --save_resume over the file chunks one by one

#### Expected behavior
ignoring linear terms for namespaces beginning with: T

#### Observed Behavior
ignoring linear terms for namespaces beginning with: T T T T T T T T T
The list keeps repeating for each save_resume call.

#### Environment
Latest master build.

### PR DESCRIPTION
**Pull request by zwd-ms**
Fix #3335

### CODE REVIEW

**Comment by jackgerrits** [comment]
When there is only a single commit `Require semantic PR title / check (pull_request_target) ` requires the commit title to also have a semantic title because of the way Github merges single commit prs. If you add another commit this requirement will go away

File: vowpalwabbit/parse_args.cc [hunk 1]
```diff
@@ -817,10 +817,11 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
 
     if (!all.logger.quiet)
     {
-      *(all.trace_message) << "ignoring namespaces beginning with: ";
-      for (auto const& ignore : ignores)
-        for (auto const character : ignore) *(all.trace_message) << character << " ";
-
+      *(all.trace_message) << "ignoring namespaces beginning with:";
+      for (size_t i = 0; i < 256; ++i)
```
[the remark this answers is not in the document]
Comment (zwd-ms) [reply vowpalwabbit/parse_args.cc]: done

**Comment by zwd-ms** [comment]
> When there is only a single commit [`Require semantic PR title / check (pull_request_target)`](https://github.com/VowpalWabbit/vowpal_wabbit/runs/4370697867?check_suite_focus=true) requires the commit title to also have a semantic title because of the way Github merges single commit prs. If you add another commit this requirement will go away

thanks!

File: vowpalwabbit/parse_args.cc [hunk 2]
```diff
@@ -859,10 +861,11 @@ void parse_feature_tweaks(options_i& options, VW::workspace& all, bool interacti
...
     if (!all.logger.quiet)
     {
-      *(all.trace_message) << "using namespaces beginning with: ";
-      for (auto const& keep : keeps)
-        for (auto const character : keep) *(all.trace_message) << character << " ";
-
+      *(all.trace_message) << "using namespaces beginning with:";
+      for (size_t i = 0; i < NUM_NAMESPACES; ++i)
+      {
+        if (!all.ignore[i]) *(all.trace_message) << " " << static_cast<unsigned char>(i);
```
Comment (cheng-tan) [inline vowpalwabbit/parse_args.cc:867]: should this be all.keeps[i]?

**Comment by jackgerrits** [comment]
It seems like this does not address the duplicated terms which are saved in command line args the model file. It just stops duplicates from being printed for the output.

**Comment by zwd-ms** [comment]
> It seems like this does not address the duplicated terms which are saved in command line args the model file. It just stops duplicates from being printed for the output.

true, let me see how to dedup in the model file.

**Review by jackgerrits — APPROVED** [review]

**Pull request merged**